### PR TITLE
feature: image loader support gamestatefulset, issue #684

### PR DESCRIPTION
**Pull request by stonewesley**

### CODE REVIEW

File: bcs-services/bcs-webhook-server/internal/plugin/imageloader/bcsgs.go [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+
+// LoadImageBeforeUpdate prevents workload instance from updating and
+// create a image-load job.
+func (b *bcsgsWorkload) LoadImageBeforeUpdate(ar v1beta1.AdmissionReview) *v1beta1.AdmissionResponse {
+	// get req gd
+	newGS := &bcsgstkexv1alpha1.GameStatefulSet{}
+	raw := ar.Request.Object.Raw
+	err := json.Unmarshal(raw, newGS)
+	if err != nil {
+		blog.Errorf("get new gd failed: %v", err)
```
Comment (abstractmj) [inline bcs-services/bcs-webhook-server/internal/plugin/imageloader/bcsgs.go:94]: gd ---> gs

File: bcs-services/bcs-webhook-server/internal/plugin/imageloader/bcsgs.go [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+	if newGS.Spec.UpdateStrategy.Type != bcsgstkexv1alpha1.InplaceUpdateGameStatefulSetStrategyType {
+		return toAdmissionResponse(nil)
+	}
+
+	// get old gd
+	oldGS := &bcsgstkexv1alpha1.GameStatefulSet{}
+	raw = ar.Request.OldObject.Raw
+	err = json.Unmarshal(raw, oldGS)
+	if err != nil {
+		blog.Errorf("get old gd failed: %v", err)
```
Comment (abstractmj) [inline bcs-services/bcs-webhook-server/internal/plugin/imageloader/bcsgs.go:108]: gd ---> gs

File: bcs-services/bcs-webhook-server/internal/plugin/imageloader/bcsgs.go [hunk 3]
```diff
@@ -0,0 +1,385 @@
...
+		// TODO find the job and delete
+	}
+	if finalJob != nil {
+		go b.i.createJob(finalJob)
+	}
+
+	return toAdmissionResponse(nil, finalPatch)
+}
+
+// imageDiff diffs old and new gd, and return patch string and containers for job to load images.
```
Comment (abstractmj) [inline bcs-services/bcs-webhook-server/internal/plugin/imageloader/bcsgs.go:194]: gd ---> gs

File: bcs-services/bcs-webhook-server/internal/plugin/imageloader/bcsgs.go [hunk 4]
```diff
@@ -0,0 +1,385 @@
...
+	return patchStr, retContainers
+}
+
+// JobDoneHook is called after image-load job is done.
+// This function trigger the update keepgoing or attachs failed event to the workload instance.
+func (b *bcsgsWorkload) JobDoneHook(namespace, name string, event *corev1.Event) error {
+	// get gd and update
+	gd, err := b.lister.GameStatefulSets(namespace).Get(name)
+	if err != nil {
+		blog.Errorf("get gd %s-%s failed: %v", namespace, name, err)
```
Comment (abstractmj) [inline bcs-services/bcs-webhook-server/internal/plugin/imageloader/bcsgs.go:242]: gd ---> gs

**Review by abstractmj — APPROVED** [review]
LGTM

**Review by abstractmj — APPROVED** [review]

**Review by abstractmj — APPROVED** [review]

**Review by DeveloperJim — APPROVED** [review]
LGTM

**Pull request merged**